### PR TITLE
Converge with openspending conductor API

### PR DESCRIPTION
**Pull request by subhankarb**
There were total 3 issues to make dpmpy converge with open-spending API.
- #84 
- #83 
- #82 

Making pull request for three separately was a problem as all three issues were closely related to each other . That is why raising the pull request for all three issues at a time.
Below are the commits for each issues:

- #84 Change request header for authorization. Addressed in https://github.com/subhankarb/dpmpy/commit/a832e258ffec8e166633e9706586ae5a9360eaee
  -  This commit fixed changes in auth header. Changes to `Auth-Token` to match gobble.
- #83 Changes the publish method. Addressed in https://github.com/subhankarb/dpmpy/commit/fa55f0dc2bea166b77652e69f4b2816938d51a17
  - Gather all file information and build request payload to fetch s3 authorized post method
  - Remove all datapackage.json upload API calls
  - Removed all unneeded testes
  - Remove datapackage.json upload API call from all tests. 
- #82 Replace finalize API with upload API https://github.com/subhankarb/dpmpy/commit/0114dd6e7319eef61779c46cbcfb2032a83dc749

### CODE REVIEW

File: dpm/client/__init__.py [hunk 1]
```diff
@@ -149,6 +154,43 @@ def publish(self, publisher=None):
...
+        local_path = join(self.datapackage.base_path, path)
+        md5 = md5_file_chunk(local_path)
+        size = getsize(local_path)
+        return {
+            'size': size,
+            'md5': md5,
+            'type': None
+        }
+
     def _upload_file(self, path, local_path):
```
Comment (rufuspollock) [inline dpm/client/__init__.py]: This should have been deleted as its functionality has merged into `_get_file_info` and and `_prepare_files_authorize`

File: dpm/client/__init__.py [hunk 2]
```diff
@@ -149,6 +154,43 @@ def publish(self, publisher=None):
         # Return published datapackage url
         return self.server + '/%s/%s' % (self.username, self.datapackage.descriptor['name'])
 
+    def _prepare_files_authorize(self):
+        files = filter(isfile, listdir(self.datapackage.base_path))
+
+        accepted_descriptor = 'datapackage.json'
```
Comment (rufuspollock) [inline dpm/client/__init__.py]: Why do we need to check through file list for `datapackage.json` - we know it must be there as this is a data package (if it is not there we should even attempt the publish ...)

File: dpm/client/__init__.py [hunk 3]
```diff
@@ -149,6 +154,43 @@ def publish(self, publisher=None):
...
+        descriptor = [f for f in files if f == accepted_descriptor][0]
+        file_list = [descriptor]
+        for resource in self.datapackage.resources:
```
Comment (rufuspollock) [inline dpm/client/__init__.py]: Blank line before this line would make this easier to read. I'd also suggest we have README before resources in the upload sequence.

File: dpm/client/__init__.py [hunk 4]
```diff
@@ -122,23 +122,28 @@ def publish(self, publisher=None):
...
-                )
+                method='POST',
+                url='/api/datastore/authorize',
+                json=self._prepare_files_authorize()
+            )
+
+        filedata = response.json().get('filedata')
+        if not filedata:
+            raise DpmException('server did not provide upload authorization for files')
+        for item in sorted(filedata.keys()):
```
Comment (rufuspollock) [inline dpm/client/__init__.py]: We want to upload in a specific order IMO:

- datapackage.json
- README
- Resources in the order they are in in the resources list in the Data Package

That is logical for the user and also means small files first.

Does this code respect that order?

File: tests/tests_cli/test_publish_invalid.py [hunk 5]
```diff
@@ -33,7 +34,7 @@ def setUp(self):
         patch('dpm.client.DataPackage', lambda *a: self.valid_dp).start()
         patch('dpm.client.exists', lambda *a: True).start()
 
-
+    @unittest.skip("INVALID NO VALIDATION CHECK FOR datapackage.json")
```
Comment (rufuspollock) [inline tests/tests_cli/test_publish_invalid.py]: Why not just delete this test if we no longer have it on the server?

**Comment by rufuspollock** [comment]
Comments on the commit messages:

* For a large commit, more detail is needed. See this example of mine. This was very large change and you would not need as much info): https://github.com/frictionlessdata/dpr-js/commit/c51668c485d91f099b11c4a06e8eed604831e58d
  * Also some typo / layout issues and is somewhat short. `-refs #xx` should be `- refs #xx` (this may seem detail but -refs may not work with gihub ref stuff ...)
* Second message was good

File: dpm/client/__init__.py [hunk 6]
```diff
@@ -272,7 +272,7 @@ def _apirequest(self, method, url, *args, **kwargs):
 
         headers = kwargs.pop('headers', {})
         if self.token:
-            headers.setdefault('Authorization', 'Bearer %s' % self.token)
+            headers.setdefault('Auth-Token', '%s' % self.token)
```
Comment (rufuspollock) [inline dpm/client/__init__.py]: Does this require **no** changes to tests?

**Comment by rufuspollock** [comment]
@subhankarb and can you write a good new summary that i can copy and paste to use when i merge (basically a summary commit message with good amount of detail).

File: dpm/client/__init__.py (hunk 6 above)
Comment (subhankarb) [inline dpm/client/__init__.py]: @rufuspollock no.

**Comment by subhankarb** [comment]
We want to make dpmpy to work with API of both [dpr-api](https://github.com/frictionlessdata/dpr-api) and [os-conductor](https://github.com/openspending/os-conductor). The epic https://github.com/frictionlessdata/dpr-api/issues/249 addressed all the changes we wanted to make so that dpmpy works with both the APIs.

In epic https://github.com/frictionlessdata/dpr-api/issues/249 we mentioned three issues that is related to dpmpy.
- #82 Modifying the `finalize` API to `publish` API. The commit https://github.com/subhankarb/dpmpy/commit/0114dd6e7319eef61779c46cbcfb2032a83dc749 contains the changes.
  - In publish command, the data package `https://example.com/api/package/%s/%s/finalize` API call replaced with `https://example.com/api/package/upload`
  - Updated relevant test cases where ever old API call has been used.

- #83 Change upload file process. The commit https://github.com/subhankarb/dpmpy/commit/fa55f0dc2bea166b77652e69f4b2816938d51a17
contains the changes.
  - Removed upload API call for `datapackage.json`
  - Added `/api/datastore/authorize` API call.
  - Changes all relevant tests.

- #84 Change request header key for jwt token. The commit https://github.com/subhankarb/dpmpy/commit/a832e258ffec8e166633e9706586ae5a9360eaee contains the changes.
  - Changes the header key for jwt token from `Authorization` to `Auth-Token` to match os-conductor.

**Comment by rufuspollock** [comment]
@subhankarb much improved - well done. I've done a detailed example that i can share of how you can improve even further!

File: dpm/client/__init__.py [hunk 7]
```diff
@@ -160,10 +160,15 @@ def publish(self, publisher=None):
...
+        data_package_s3_url = filedata['datapackage.json']['upload_url']
         response = self._apirequest(
             method='POST',
-            url='/api/package/%s/%s/finalize' % (self.username, self.datapackage.descriptor['name'])
+            url='/api/package/upload',
+            json={'datapackage': data_package_s3_url}
         )
+        status = response.json().get('status', None)
+        if status is None or status != 'queued':
+            raise DpmException('server did not provide upload authorization for files')
```
Comment (rufuspollock) [inline dpm/client/__init__.py]: Strictly i think the error message is something like: "Unable to finalize publication of the data package. Data is in bitstore but not published."

However for now I think this is ok.

**Pull request merged**